### PR TITLE
[DSpark] Capture the full draft step as a decode CUDA graph 

### PR DESCRIPTION
**Pull request by Juntian777**
## Purpose

The DFlash/DSpark draft step keeps two pieces of per-step host work on the decode hot path, even when the query forward replays a FULL CUDA graph:

1. The context K/V precompute (`precompute_and_store_context_kv`) runs eagerly outside the graph because the context shape varies per step.
2. The attention metadata build runs eagerly every step purely for its side effect of refreshing the SWA builder's persistent GPU buffers — the built metadata itself is discarded on the replay path.

On uniform decode, both are fixed-shape and capture-safe, so this PR captures the entire draft step — context K/V precompute, metadata refresh, query forward, and Markov sampling — as a single decode-step CUDA graph.

**Before this PR:**
image

**After this PR:**
bdf45551724c45e395cf8a764dae2123

## Test plan

Capturing the draft step must not change what the speculator proposes, so the
check is an end-to-end accuracy A/B against this branch's parent commit with
everything else held fixed.

GSM8K, 5-shot, temperature 0, concurrency 32, 1319 problems, on
DeepSeek-V4.1-Flash served with DSpark speculative decoding. 4x GB200, DEP4
(DP4 x TP1, experts over EP)

```bash
vllm serve deepseek-ai/DeepSeek-V4.1-Flash \
  --data-parallel-size 4 --enable-expert-parallel \
  --reasoning-parser deepseek_v41 \
  --kernel-config '{"moe_backend":"deep_gemm_mega_moe"}' \
  --attention-config '{"backend":"FLASHMLA_MEGA_ATTN_DSV41","indexer_kv_dtype":"mxfp4","indexer_sparse_logits":true}' \
  --kv-cache-dtype fp8 \
  --max-model-len 8192 --max-num-seqs 128 \
  --speculative-config '{"method":"dspark","num_speculative_tokens":5,"enable_adaptive_verification":true}'
```

## Test result

| Build | GSM8K exact match (strict) |
| --- | ---: |
| This PR (`c17018b1c8`) | **0.9227** |
| Parent (`c723a831a8`) | **0.9249** |

The gap is 0.22 pp, or 3 of 1319 problems, against a per-run standard error of
roughly 0.7 pp — the captured draft step leaves the sampled output unchanged.

AI assistance was used for this change: the implementation was written with
Claude Code

### CODE REVIEW

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 1]
```diff
@@ -145,22 +149,95 @@ def init_cudagraph_manager(self, cudagraph_mode: CUDAGraphMode) -> None:
...
+            and self.query_cudagraph_manager.needs_capture()
+            and self.pcp_manager is None
+            and self.block_tables.cp_size == 1
+        ):
+            self.draft_step_cudagraph_manager = CudaGraphManager(
+                self.vllm_config,
+                self.device,
+                cudagraph_mode,
+                decode_query_len=self.num_query_per_req,
+            )
```
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: Instead of creating a separate cudagraph manager for the fused version, can we instead extend `DFlashCudaGraphManager` to optionally perform the `_precompute_context_kv` operation?

**Review by TheEpicDolphin — CHANGES_REQUESTED** [review]
Could you please run some before & after benchmarks to see what we gain from these channges?

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py (hunk 1 above)
Comment (Juntian777) [reply vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: Thanks for the review! I've updated my code to extend DFlashCudaGraphManager and added a performance benchmark.

**Review by TheEpicDolphin — CHANGES_REQUESTED** [review]
Thanks for updating the PR, this looks much cleaner. After thinking about it for some time, I'd prefer if we only ship the `precompute_context_kv` fusion into the cudagraph, and not the `build_attn()`. I think that the ability to fuse the attention metadata build into the cudagraph should be a property of the attention backends, not the model. 

We already have a pattern for fusing metadata builds into the cudagraph. It's done here for MTP/EAGLE: https://github.com/vllm-project/vllm/blob/711fc55c10a2bbbefcb12456c43ba5eca01be3f6/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L637-L638
Basically, we enable fusion whenever that method is implemented by all of the draft model's attention backends. It's intended to be lightweight, cudagraph-capture safe, and only for decodes. It enables drafting the last N-1 tokens without needing to rebuild the metadata each time.

I'd like to bring that pattern here for DFlash/DSpark, but as a separate effort. Some extra changes are needed for some of DSV4.1's backends to enable this. I'm planning to work on this soon, but if the optimization is needed urgently I'm happy to share the details of what's needed.

In summary, can you please strip the attention metadata build from the cudagraph, and just fuse `precompute_context_kv`? That should simplify things further, while still keeping most of the performance gains.

**Comment by Juntian777** [comment]
> Thanks for updating the PR, this looks much cleaner. After thinking about it for some time, I'd prefer if we only ship the `precompute_context_kv` fusion into the cudagraph, and not the `build_attn()`. I think that the ability to fuse the attention metadata build into the cudagraph should be a property of the attention backends, not the model.
> 
> We already have a pattern for fusing metadata builds into the cudagraph. It's done here for MTP/EAGLE:
> 
> https://github.com/vllm-project/vllm/blob/711fc55c10a2bbbefcb12456c43ba5eca01be3f6/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L637-L638
> 
> 
> Basically, we enable fusion whenever that method is implemented by all of the draft model's attention backends. It's intended to be lightweight, cudagraph-capture safe, and only for decodes. It enables drafting the last N-1 tokens without needing to rebuild the metadata each time.
> I'd like to bring that pattern here for DFlash/DSpark, but as a separate effort. Some extra changes are needed for some of DSV4.1's backends to enable this. I'm planning to work on this soon, but if the optimization is needed urgently I'm happy to share the details of what's needed.
> 
> In summary, can you please strip the attention metadata build from the cudagraph, and just fuse `precompute_context_kv`? That should simplify things further, while still keeping most of the performance gains.

@TheEpicDolphin  Thanks for the advice. I agree that attention metadata cudagraph fusion should live in the attention backend. I’ve updated the PR to keep only precompute_context_kv in the cudagraph now and removed build_attn(). I’ve also updated the benchmark results accordingly.

**Review by TheEpicDolphin — APPROVED** [review]
This looks good to me, thank you! Please run CI on this PR when you get the chance.

**Review by zyongye — APPROVED** [review]
